src/middlewared/middlewared/utils/nss/grp.py

```diff
@@ -0,0 +1,245 @@
+    __setgrent(mod)
+    group_list = []
+
+    group = __getgrent_impl(mod, as_dict)
+    while group is not None:
+        if (group := __getgrent_impl(mod, as_dict)):
+            group_list.append(group)
+
+    __endgrent(mod)
+    return group_list
+
+
+def __getgrnam_impl(name, module, as_dict):
+    mod = NssModule[module]
+    result = Group()
+    buf = ctypes.create_string_buffer(GROUP_BUFLEN)
+
+    res, errno, result_p = __getgrnam_r(name, ctypes.byref(result),
+                                        buf, GROUP_BUFLEN, mod)
+    if errno != 0:
+        raise NssError(errno, NssOperation.GETGRNAM, res, mod)
+
+    if res == NssReturnCode.NOTFOUND:
+        return None
+
+    return  __parse_nss_result(result, as_dict, mod.name)
+
+
+def __getgrgid_impl(gid, module, as_dict):
+    mod = NssModule[module]
+    result = Group()
+    buf = ctypes.create_string_buffer(GROUP_BUFLEN)
+
+    res, errno, result_p = __getgrgid_r(gid, ctypes.byref(result),
+                                        buf, GROUP_BUFLEN, mod)
+    if errno != 0:
+        raise NssError(errno, NssOperation.GETGRGID, res, mod)
+
+    if res == NssReturnCode.NOTFOUND:
+        return None
+
+    return  __parse_nss_result(result, as_dict, mod.name)
+
+
+def getgrgid(gid, module=NssModule.ALL.name, as_dict=False):
+    if module != NssModule.ALL.name:
+        if (result := __getgrgid_impl(gid, module, as_dict)):
+            return result
+
+        raise KeyError(f"getgrgid(): gid not found: '{gid}'")
+
+    # We're querying all modules
+    for mod in NssModule:
+        if mod == NssModule.ALL:
+            continue
+
+        try:
+            if (result := __getgrgid_impl(gid, mod.name, as_dict)):
+                return result
+        except NssError as e:
+            if e.return_code != NssReturnCode.UNAVAIL:
+                raise e from None
+
+    raise KeyError(f"getgrgid(): gid not found: '{gid}'")
+
+
+def getgrnam(name, module=NssModule.ALL.name, as_dict=False):
+    if module != NssModule.ALL.name:
+        if (result := __getgrnam_impl(name, module, as_dict)):
+            return result
+
+        raise KeyError(f"getgrnam(): name not found: '{name}'")
+
+    # We're querying all modules
+    for mod in NssModule:
+        if mod == NssModule.ALL:
+            continue
+
+        try:
+            if (result := __getgrnam_impl(name, mod.name, as_dict)):
+                return result
+        except NssError as e:
+            if e.return_code != NssReturnCode.UNAVAIL:
+                raise e from None
+
+    raise KeyError(f"getgrnam(): name not found: '{name}'")
+
+
+def getgrall(module=NssModule.ALL.name, as_dict=False):
+    if module != NssModule.ALL.name:
+        return {module: __getgrall_impl(module, as_dict)}
+
+    results = {}
+    for mod in NssModule:
+        if mod == NssModule.ALL:
+            continue
+
+        try:
+            if (entries := __getgrall_impl(mod.name, as_dict)):
+                results[mod.name] = entries
```

Comment:
Same comment as for pwd.  Although, since there's no comment on `getgrall` we can't tell here what the intended behavior is! :-)